modules/db/src/main/scala/gem/dao/EphemerisDao.scala

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2016-2017 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package gem
+package dao
+
+import gem.enum.EphemerisKeyType
+import gem.math._
+import gem.util.InstantMicros
+
+import cats.Monad
+import cats.implicits._
+import doobie._, doobie.implicits._
+import fs2.Stream
+
+object EphemerisDao {
+
+  def insert(k: EphemerisKey, e: Ephemeris): ConnectionIO[Int] =
+    Statements.insert.updateMany(
+      e.toMap.toList.map { case (i, c) => (k, i, c, c.ra.format, c.dec.format) }
+    )
+
+  def streamInsert[M[_]: Monad](k: EphemerisKey, s: Stream[M, Ephemeris.Element], xa: Transactor[M]): Stream[M, Int] =
+    Stream.constant(k)                                                  // Stream[Pure, EphemerisKey]
+      .zip(s)                                                           // Stream[M, (EphemerisKey, Ephemeris.Element)]
+      .map { case (k, (i, c)) => (k, i, c, c.ra.format, c.dec.format) } // Stream[M, EphemerisRow]
+      .segmentN(4096)                                                   // Stream[M, Segment[EphemerisRow, Unit]]
+      .flatMap { rows =>
+        Stream.eval(Statements.insert.updateMany(rows.toVector).transact(xa))
+      }
+
+  def delete(k: EphemerisKey): ConnectionIO[Int] =
+    Statements.delete(k).run
+
+  def update(k: EphemerisKey, e: Ephemeris): ConnectionIO[Unit] =
+    (delete(k) *> insert(k, e)).void
```

Comment:
i don't know enough doobie, so this is more a question than a comment. Is this usual? to delete and insert rather than do a straight update?

Reply:
This is intended to support a change in the horizons output where we want to wholesale replace all the ephemeris element rows.  The rows are logically keyed by the id and timestamp and the timestamps themselves are potentially being updated with new horizons output so it seemed like the best option to just wipe them out and write the new data.

Now that I'm looking at this again though, I probably should have actually made an actual primary key out of the triple (`key_type`, `key`, `timestamp`).  I'll go back and do that.

Reply:
We might also want `streamUpdate` here. I guess we'll see how it performs.